Added gltf-pipeline command line interface

JoshuaStorm opened the pull request:
Added gltf-pipeline command line interface.
Refactored pipeline to separate file writing and optimizations. 
Added gltfPipelineSpec tests.

Pair programming with @likangning93

pjcozzi commented:
JSHint is failing: https://travis-ci.org/AnalyticalGraphicsInc/gltf-pipeline/builds/134604710

pjcozzi commented:
@lilleyse do you want to do a first review before I look at this?

@leerichard42 are you also interested in looking at this?

lilleyse commented:
> @lilleyse do you want to do a first review before I look at this?

Sure.

likangning93 commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,106 @@
+        }
+        else if (fileExtension === '.gltf') {
+            gltf = JSON.parse(data);
+            addPipelineExtras(gltf);
```
We moved `addPipelineExtras` here from further down to resolve a bug where the gltfPipeline couldn't load glbs with the `KHR_binary_glTF extension`. It looks like glbs with that extension have an empty data URI, which isn't a problem for the pipeline if the data that should be there is in the assets' `extra` attribute. `parseBinaryGltf` was handling this, but then the later call to `addPipelineExtras` seemed to wipe out the `extra` attributes, causing an error later down in `loadGltfUris`.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
+var defaultValue = Cesium.defaultValue;
 var defined = Cesium.defined;
+var writeGltf = require('../').writeGltf;
+var writeBinaryGltf = require('../').writeBinaryGltf;
```
Change these two includes to be like the new ones: `require('../lib/writeGltf');` (unless there was a good reason to keep it the original way)

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
-	var help =
+    var help =
         'Usage: node ' + path.basename(__filename) + ' [path-to.gltf or path-to.bgltf] [OPTIONS]\n' +
-        '  -o=PATH  Write optimized glTF to the specified file.\n';
```
Put each of the options of their own line. For organization purposes, -o should go after -i.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
         'Usage: node ' + path.basename(__filename) + ' [path-to.gltf or path-to.bgltf] [OPTIONS]\n' +
-        '  -o=PATH  Write optimized glTF to the specified file.\n';
+        '  -i, input=PATH Read unoptimized glTF from the specified file.\n  -b, write binary glTF file.\n' +
+        '  -s, output non-embedded files.\n  -o, output=PATH write optimized glTF to the specified file.\n';
```
Change the -s description, maybe like this from ModelConverter:

'writes out separate geometry/animation data files, shader files and textures instead of embedding them in the glTF file.'

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
-        if (err) {
-            throw err;
-        }
+var outputPath = argv.o;
```
Also worth mirroring -i with something like:

`var outputPath = defaultValue(argv._[1], argv.o);`

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
+    if (exportBinary) {
+        var outputExtension = path.extname(outputPath);
+        if (outputExtension !== ".glb") {
+            outputPath = path.basename(outputPath, outputExtension) + ".glb";
```
Usually we use single quotes around strings.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
+        var outputExtension = path.extname(outputPath);
+        if (outputExtension !== ".glb") {
+            outputPath = path.basename(outputPath, outputExtension) + ".glb";
         }
```
You can move this check to `writeBinaryGltf`.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
+        if (outputExtension !== ".glb") {
+            outputPath = path.basename(outputPath, outputExtension) + ".glb";
         }
+        writeBinaryGltf(gltf, outputPath, true);
```
Note: we'll also want to pass `isSeparate` in here too soon.

lilleyse commented on index.js:
```diff
@@ -26,5 +26,6 @@ module.exports = {
     removeUnusedVertices : require('./lib/removeUnusedVertices'),
-    OptimizationStatistics : require('./lib/OptimizationStatistics')
+    OptimizationStatistics : require('./lib/OptimizationStatistics'),
+    gltfPipeline : require('./lib/gltfPipeline')
```
I'm starting to wonder if `optimizeGltf` is a better name here now that it doesn't input and output files.

pjcozzi replied:
Keep in mind that it will have stages to do more than optimization, e.g., add AO.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,106 @@
+var removeUnusedMeshes = require('../').removeUnusedMeshes;
+var removeUnusedNodes = require('../').removeUnusedNodes;
+var removeUnusedAccessors = require('../').removeUnusedAccessors;
+var removeUnused = require('./removeUnused');
```
Same comment here making the includes use the same style.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,106 @@
+var defaultValue = Cesium.defaultValue;
+var DeveloperError = Cesium.DeveloperError;
+
+
```
Remove extra line.

lilleyse commented on lib/writeBinaryGltf.js:
```diff
@@ -106,6 +106,8 @@ function writeBinaryGltf(gltf, outputPath, createDirectory, callback) {
     var scene = new Buffer(sceneString);
     var glb = Buffer.concat([header, scene, body], glbLength);
 
+    console.log(outputPath);
```
Remove this.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,30 @@
+var glbPath = './specs/data/boxTexturedUnoptimized/CesiumTexturedBoxTest.glb';
+
+describe('gltfPipeline', function() {
+    it('checks that gltf is defined when passing in a gltf file', function(done) {
```
Make the test name a little more descriptive, like `optimizes a gltf file`.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,30 @@
+            inputPath : gltfPath,
+            separate : false
+        }, function(gltf) {
+            expect(defined(gltf)).toEqual(true);
```
Use `.toBeDefined()` instead. Also is there any way to make this more thorough? Maybe at least check that the gltfs don't equal each other.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,30 @@
+        });
+    });
+
+    it('checks that gltf is defined when passing in a glb file', function(done) {
```
And here: "optimized a binary glTF file"

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,106 @@
+
+module.exports = gltfPipeline;
+
+function gltfPipeline(options, callback) {
```
Since this will be the starting point for projects that use gltf-pipeline as a library, we should also accept gltf objects directly. Definitely the obj2gltf will use it this way instead of writing a file first, then calling this. Maybe there should be another file that reads gltf files? @pjcozzi what do you think?

pjcozzi replied:
Yes, exactly.  At the top-level there should be the command-line app, which just calls a function like this, which then calls a function that takes glTF in-memory as input.

Same deal for output - one function would just provide glTF in-memory and another would write to disk.

So the complete Node.js server -> obj2gltf -> gltf-pipeline flow will not write to disk.

Note that the challenge here is to define the in-memory data structure, i.e., it's one thing for the JSON to be loaded, but what about the .bin, .glsl, and images?  Perhaps gltf-pipeline should document its simple in-memory structure and provide helpers to read and write it (which it already does).

JoshuaStorm commented:
Updated, addressed @lilleyse's comment, aside from last (awaiting @pjcozzi's input).

Currently working to add more gltfPipelineSpec tests.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
 #!/usr/bin/env node
 'use strict';
 var fs = require('fs');
+var argv = require('yargs').argv;
```
Add yargs to package.json, and remove minimist.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,53 @@
+var writeBinaryGltf = require('../').writeBinaryGltf;
+var gltfPipeline = require('../lib/gltfPipeline');
 
 if (!defined(argv._[0]) || defined(argv.h) || defined(argv.help)) {
```
Now that we support -i for input files, `argv._[0]` may be undefined but shouldn't go the the help screen. As an alternative you can check for `process.argv.length < 3`.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,51 @@
+var writeBinaryGltf = require('./writeBinaryGltf');
+var gltfPipeline = require('../lib/gltfPipeline');
 
 if (!defined(argv._[0]) || defined(argv.h) || defined(argv.help)) {
```
Adding this comment again because the previous comment is hidden:

Now that we support -i for input files, argv._[0] may be undefined but shouldn't go the the help screen. As an alternative you can check for process.argv.length < 3.

leerichard42 commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,106 @@
+            combinePrimitives(gltf);
+
+            removeUnusedVertices(gltf);
+
```
I was actually thinking about moving removeUnusedVertices to come before the combine meshes/primitives/nodes stage, so that the actual scene data would be optimized before condensing the gltf structure.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,106 @@
+    var filePath = path.dirname(gltfPath);
+
+    if (fileExtension !== '.glb' && fileExtension !== '.gltf') {
+        throw new Error('Invalid glTF file.');
```
To be consistent this should be `DeveloperError`.

JoshuaStorm commented:
Updated. 
- [ ] Still need to cleanup callbacks in gltf-pipeline and loadGltfUris

pjcozzi commented:
@lilleyse let me know when this is ready for me to review.

likangning93 commented:
Updated!

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,55 @@
+var gltfPipeline = require('../lib/gltfPipeline');
+var addPipelineExtras = require('../lib/addPipelineExtras');
+var readGltf = require('../lib/readGltf');
+var DeveloperError = Cesium.DeveloperError;
```
Put this include near the other Cesium ones.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,55 @@
+        '  -i, input=PATH Read unoptimized glTF from the specified file.\n ' +
+        '  -b, write binary glTF file.\n' +
+        '  -s, writes out separate geometry/animation data files, shader files and textures instead of embedding them in the glTF file.\n  ' +
+        '  -o, output=PATH write optimized glTF to the specified file.\n';
```
Put the -o line after the -i.

lilleyse commented on bin/gltf-pipeline.js:
```diff
@@ -1,114 +1,55 @@
-    process.stdout.write('Samplers removed: ' + stats.numberRemoved.samplers + '\n');
-    process.stdout.write('Shaders removed: ' + stats.numberRemoved.shaders + '\n');
-}
+});
```
Not a big deal, but add a newline here. And wherever else you see it.

lilleyse commented on lib/loadGltfUris.js:
```diff
@@ -15,21 +15,11 @@ function loadGltfUris(gltf, basePath, callback) {
     async.each(['buffers', 'images', 'shaders'], function(name, asyncCallback) {
             loadURI(gltf, basePath, name, asyncCallback);
         }, function(err) {
```
There's some extra indentation in this area.

lilleyse commented on lib/loadGltfUris.js:
```diff
@@ -15,21 +15,11 @@ function loadGltfUris(gltf, basePath, callback) {
-            }
-            else if (callback) {
-                callback();
+            if (defined(callback)) {
```
The callback should always be defined, right?

likangning93 replied:
`combinePrimitivesSpec` has a couple specs that run `loadGltfUris` without a callback. I'll see if I can change these.

lilleyse commented on lib/readGltf.js:
```diff
@@ -0,0 +1,47 @@
+var fs = require('fs');
+var parseBinaryGltf = require('./parseBinaryGltf');
+var addPipelineExtras = require('./addPipelineExtras');
+var loadGltfUris = require('../lib/loadGltfUris');
```
Simplify to `var loadGltfUris = require('./loadGltfUris');`

lilleyse commented on lib/readGltf.js:
```diff
@@ -0,0 +1,47 @@
+
+    var fileExtension = path.extname(gltfPath);
+    var filePath = path.dirname(gltfPath);
+    
```
Remove extra line.

lilleyse commented on lib/readGltf.js:
```diff
@@ -0,0 +1,47 @@
+            callback(gltf);
+        });
+    });
+}
```
Newline.

lilleyse commented on lib/writeBinaryGltf.js:
```diff
@@ -15,13 +15,19 @@ var removePipelineExtras = require('./removePipelineExtras');
     }
 
-    //Create the special binary buffer from the existing buffers
+
```
Remove extra line.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,33 @@
+var glbPath = './specs/data/boxTexturedUnoptimized/CesiumTexturedBoxTest.glb';
+
+describe('gltfPipeline', function() {
+    it('optimizes a gltf file, checks that a new gltf is defined when passing in a gltf file', function(done) {
```
For this title, just `optimizes a gltf file` should be enough. Same comment below.

lilleyse commented on specs/lib/readGltfSpec.js:
```diff
@@ -0,0 +1,54 @@
+                expect(err.message).toEqual('Input path is undefined.');
+                throw err;
+            }
+        }).toThrow();
```
These two tests can be simplified, for example this one could just be:

```
expect(function() {
    readGltf(undefined);
}).toThrow();
```

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,33 @@
+describe('gltfPipeline', function() {
+    it('optimizes a gltf file', function(done) {
+        readGltf(gltfPath, function(gltf) {
+            var stringCopy = JSON.stringify(gltf);
```
Now that I've looked more carefully, we should just use the `clone` library like some other tests do. Look at `writeShaderSpec` for example.

lilleyse commented:
@pjcozzi You can review this whenever you're ready.

lilleyse commented:
Actually there are still a few spec changes in progress, I'll update when they're ready.

likangning93 commented:
I pushed some changes to the combinePrimitivesSpec. Some documentation on what happened:

We found out by changing `expect` to `expect.not` to see if all the tests were actually running properly, which revealed the `expect`s in `does not affect single primitives` and `does not combine two primitives` were never getting hit. This seemed to be due to an error in calls to `writeGltf`, which were preventing the callbacks containing the `expect`s to run. Aside: @leerichard42 , what was the `writeGltf` here for?
Figuring this out was obfuscated heavily by other tests failing with bizarre outputs, like `expect`s receiving whole gltf objects instead of the small values they were supposed to be comparing with. We suspect what might have happened was that jasmine somehow blocks file output and leaves those unwritten gltfs floating around somewhere that they can interfere with the other tests.

We also found that using `clone` here, including `Cesium`'s recursive clone, would run into un-clonable objects, hence the use of `stringify` in some of the updated tests.

leerichard42 commented:
The writeGltf calls in the spec were intended to return and write out a gltf object with the extras removed and data uris written out, so that I would be able to visually confirm the file in the converter.

JoshuaStorm commented:
Updated. Refactored gltfPipeline to follow @pjcozzi's guidelines more closely, my still need some work.
Changed index.js to only expose those specific files. 

Changed some require syntax in a few specs that were creating issues with Jasmine.

likangning93 commented:
Small update: I added the `options` and `callback` parameters back into the functions in  `gltfPipeline` as per @lilleyse's request

pjcozzi commented:
What's the latest on this?  Do you need me to review yet?

lilleyse commented:
I still need to do a review of the latest changes, I'll tell you when I'm ready.

lilleyse commented on index.js:
```diff
@@ -1,30 +1,9 @@
-    removeUnusedVertices : require('./lib/removeUnusedVertices'),
-    OptimizationStatistics : require('./lib/OptimizationStatistics')
+    writeGltf : require('./lib/writeGltf'),
+    writeSource : require('./lib/writeSource')
```
Should this be exposed here?

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,108 @@
+    processFile : processFile,
+    processJSONToDisk : processJSONToDisk,
+    processFileToDisk : processFileToDisk
+};
```
In order to handle the case where the user has a gltf json in memory but not with the extras, `processJSONToDisk` and `processJSON` should take in `gltf` instead of `gltfWithExtras`. They can add the extras to the gltf then call a function named `processJSONWithExtras` which actually does the pipeline steps (like what `processJSON` does currently). The new `processJson` could also remove pipeline extras before sending back the gltf.

JoshuaStorm replied:
With `ProcessJSON` should we just assume that the gltf URIs have already been loaded into the gltf file? If not, we would need an input path as an argument, sort of defeating one purpose of using a json from memory.

lilleyse replied:
Ah.. I forgot about that. Maybe options could optionally include a `resourcePath`?

JoshuaStorm replied:
Do you mean to optionally include a `resourcePath`, but otherwise assume that the URIs are already loaded?

lilleyse replied:
Yup that sounds fine to me.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,108 @@
+    processFileToDisk : processFileToDisk
+};
+
+function processJSON(gltfWithExtras, options, callback) {
```
Maybe one option to start out with here is `printStats`, which is `false` by default. Probably not worth adding to the command-line options.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,108 @@
+            writeBinaryGltf(gltf, outputPath, true, callback);
+        } else {
+            writeGltf(gltf, outputPath, isEmbedded, true, callback);
+        }
```
A lot of the code in `processFileToDisk` and `processJSONToDisk` can be factored out into a common function.

lilleyse commented on lib/readGltf.js:
```diff
@@ -0,0 +1,46 @@
+    var fileExtension = path.extname(gltfPath);
+    var filePath = path.dirname(gltfPath);
+
+    if (!defined(gltfPath)) {
```
This check should go first.

lilleyse commented on specs/lib/combinePrimitivesSpec.js:
```diff
@@ -1,99 +1,84 @@
                 "type": "SCALAR",
-                "count": 516
+                "count": 516,
+                "extras": { "_pipeline": { "deleteExtras": true } }
```
Is there a way to get around checking this, it just seems out of place. Maybe a call to `removePipelineExtras` beforehand?

lilleyse replied:
Handle this comment as well, and the one below.

lilleyse commented on specs/lib/combinePrimitivesSpec.js:
```diff
@@ -102,15 +87,17 @@ describe('combinePrimitives', function() {
                 "max": [0.5, 0.5, 0.5],
-                "min": [-0.5, -0.5, -0.5]
+                "min": [-0.5, -0.5, -0.5],
+                "extras": { "_pipeline": { "deleteExtras": true } }
```
Same here.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,48 @@
+'use strict';
+var clone = require('clone');
```
This reminds me that there are still other areas that use `Cesium.clone`. Those should probably switch over to use this library as well.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,48 @@
+                done();
+            });
+        });
+        processFile(gltfPath, options, function(gltf) {
```
Make this second check a different test.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,48 @@
+                done();
+            });
+        });
+        processFile(glbPath, options, function(gltf) {
```
Also make this a separate test.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,48 @@
+            done();
+        });
+    });
+});
```
This file tests `processFile`, `processJson`, but we should also test all the other functions in `gltfPipeline`. Also in order to test the write function (json and binary), we can spy on `fs.writeFile` and check that the outputPath argument is correct. (Check out this for an example: https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Scene/ShadowMapSpec.js#L948-L967, but don't use `callThrough` because we don't want to save the file).

lilleyse commented on specs/lib/readGltfSpec.js:
```diff
@@ -0,0 +1,36 @@
+        expect(function() {
+            readGltf(undefined);
+        }).toThrow();
+        done();
```
I think it's ok to remove `done()` if the test isn't asynchronous. Also remove the `done` argument.

JoshuaStorm replied:
I thought that were the case, but in these two specific tests they timeout if the `done()` is not there. I'm not exactly sure why.

lilleyse replied:
If you remove the `done` argument it should be fine.

lilleyse commented on specs/lib/readGltfSpec.js:
```diff
@@ -0,0 +1,36 @@
+        expect(function() {
+            readGltf(invalidPath);
+        }).toThrow();
+        done();
```
Same comment here.

lilleyse commented:
If anyone has feedback on the names `processJSON`, `processFile`, `processJSONToDisk`, and `processFileToDisk` that would be helpful. I think they are pretty good, but could possibly be better.

JoshuaStorm commented:
Updated!

lilleyse commented on index.js:
```diff
@@ -1,30 +1,9 @@
-};
\ No newline at end of file
+    writeGltf : require('./lib/writeGltf'),
+    writeSource : require('./lib/writeSource')
```
I think my comment from this got lost, but do we need `writeSource` in here?

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,130 @@
+    processJSONToDisk : processJSONToDisk,
+    processFileToDisk : processFileToDisk,
+    writeFile : writeFile
+};
```
If possible we should try not to expose `processJSONWithExtras` and `writeFile` since they are really just internal to this file. Some tests may have to adjust for this.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,130 @@
+            }
+            processJSONWithExtras(gltf, options, function(gltf) {
+                callback(gltf);
+            });
```
You can simplify this to `processJSONWithExtras(gltf, options, callback)`

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,130 @@
+        });
+    } else {
+        // If a resourcePath is not passed in, assume JSON is already embedded
+        processJSONWithExtras(gltf, options, function(gltf) {
```
And here too.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+
```
Remove extra line.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,130 @@
+    }
+}
+
+function processJSONToDisk(gltf, outputPath, options, callback) {// isBinary, isEmbedded) {
```
It's fine to remove the commented code here.

lilleyse commented on lib/gltfPipeline.js:
```diff
@@ -0,0 +1,130 @@
+    });
+}
+
+function processFileToDisk(inputPath, outputPath, options, callback) {//isBinary, isEmbedded) {
```
And here.

lilleyse commented on lib/readGltf.js:
```diff
@@ -0,0 +1,45 @@
+
+    if (!defined(gltfPath)) {
+        throw new DeveloperError('Input path is undefined.');
+    }
```
Put this check above since `fileExtension` and `filePath` expect it to be defined.

lilleyse commented on lib/writeGltf.js:
```diff
@@ -1,5 +1,5 @@
 'use strict';
-var fs = require('fs');
+var fs = require('fs'); 
```
Extra space here.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+var outputPath = './output/';
+
+describe('gltfPipeline', function() {
+    it('optimizes a gltf JSON given an embedded JSON', function() {
```
A clearer description might be `optimizes a gltf JSON with embedded resources`.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+        });
+    });
+    
+    it('optimizes a gltf JSON given a resourcePath', function() {
```
And here `optimizes a gltf JSON with external resources`.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+        var gltf;
+        var gltfCopy;
+        var options = { 'resourcePath' : path.dirname(gltfPath) };
+        fs.readFile(gltfPath, function(err, data) {
```
Why not use `readGltf` here?

JoshuaStorm replied:
readGltf includes the `loadGltfUris' step, and this test is checking that `processJSON` properly handling a JSON that has yet to have that step applied

lilleyse replied:
Ah right.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+                expect(gltf).not.toBe(gltfCopy);
+            });
+        });
+    });
```
You could probably drop these two tests (related to my other comment).

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+        var gltfCopy;
+        var options = {};
+        readGltf(gltfPath, function(gltf) {
+            gltfCopy = gltf;
```
Not actually doing a clone here.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+        var gltfCopy;
+        var options = {};
+        readGltf(glbPath, function(gltf) {
+            gltfCopy = gltf;
```
And here.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+                expect(gltf).not.toBe(gltfCopy);
+            });
+        });
+    });
```
All of these tests should have `done` callbacks.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+            expect(spy).toHaveBeenCalled();
+            done();
+        });
+    });
```
These two tests can be merged into one. Don't forget to check that the expected output path.

JoshuaStorm replied:
I had an issue with trying to check for an expected output path. Because multiple files are written, it is difficult to test exactly what the output path would be (or at least I am not sure how I would go about it). 

For example: 
I tried `expect(spy).toHaveBeenCalledWith(outputPath, jasmine.any(String), jasmine.any(Function));` but the result ends up being

 `[ outputPath/output.gltf, <a stringified JSON blob>, <a function>], [ outputPath/output0VS.glsl, <a stringified JSON>, <a function>], ... `

Overall, I am just unsure how I can write a Jasmine expect for a spy that seems to see multiple outputs with different expected files. Maybe something with regular expressions?

lilleyse replied:
Oh ok, it may help if you set the `isEmbedded` options to true. Or `isEmbedded` should default to true instead of false..

JoshuaStorm replied:
That is a way easier solution, thanks!

JoshuaStorm replied:
I am having issues with merging these two tests into one where it is throwing: `Error: EISDIR: illegal operation on a directory, open 'output/output'
    at Error (native)`

I initially thought this was because it was trying to create a new directory in a path that already exists, but even after adding an option to not create a new directory I am still getting the same error.

Any thoughts? Should I just leave it as separate tests?

lilleyse replied:
I'm tried calling just the second one and it passes for me.

lilleyse replied:
The new changes look good, just check this last thing and it will be ready.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,135 @@
+            });
+        });
+    });
+});
```
Add newline.

lilleyse commented on specs/lib/readGltfSpec.js:
```diff
@@ -0,0 +1,32 @@
+        readGltf(glbPath, function(gltf) {
+            expect(gltf).toBeDefined();
+        });
+    });
```
You still need the `done` callback for these two tests.

JoshuaStorm commented:
Updated.

I believe I addressed all the comments, aside from the one about combining the two file writing tests in gltfPipelineSpec (see comment above).

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -0,0 +1,114 @@
+            gltfCopy = clone(gltf);
+            processFile(glbPath, options, function (gltf) {
+                expect(gltf).toBeDefined();
+                expect(gltf).not.toBe(gltfCopy);
```
I missed these before, but it should be `.not.toEqual` for all these tests. (`.toBe` checks that they are the same object, `.toEqual` checks values, like `===` vs `==`).

JoshuaStorm replied:
That explains why they weren't failing when I was forgetting to clone them. Thanks for the explanation

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -112,24 +92,23 @@ describe('gltfPipeline', function() {
+        var options = { 'createDirectory' : false };
         processFileToDisk(gltfPath, outputPath, options, function() {
-            expect(spy).toHaveBeenCalled();
+            expect(spy).toHaveBeenCalledWith('./output/', jasmine.any(String), jasmine.any(Function));
```
Cleaner with something like `expect(spy.calls.first().args[0]).toEqual('./output/');`

lilleyse commented:
Just those few comments, this is getting close.

JoshuaStorm commented:
Updated

JoshuaStorm commented:
Addressed those last two comments about the `gltfPipelineSpec` and `combinePrimitivesSpec`. Am I missing anything?

lilleyse commented:
I think that's everything, I'm going to merge and if we missed anything we can make another PR.

pjcozzi commented:
This is a good start for the command-line interface; is the ability to configure custom pipelines via command-line, JSON input, and a function taking a JSON object coming as a separate PR?

JoshuaStorm commented:
Unless I am misunderstanding, we currently have functions to take a JSON as input in the `gltfPipeline.js` file --`processJSON` and `processJSONToDisk`.

Also, could you elaborate a bit more on what you mean by "ability to configure custom pipelines via command-line"

Thanks!

pjcozzi commented:
From #1:

> * [ ] Command-line and JSON input, e.g., like pdal (see [here](http://www.pdal.io/apps.html) and [here](http://www.pdal.io/pipeline.html))

lasalvavida commented:
@JoshuaStorm I could be wrong, but I think the idea is that the pipeline should be able to output to stdout, read from stdin, and specify specific operations to perform. Then you could chain commands together using pipes. Something like:
`./bin/gltf-pipeline.js --stdout myModel.gltf --stage addDefaults | ./bin/gltf-pipeline.js --stdin --stage quantizeAttributes`
